standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/DirectSqlUpdatePart.java

```diff
@@ -0,0 +1,1536 @@
+
+      if (insertMap.size() != 0) {
+        insertIntoPartColStatTable(insertMap, csId, dbConn);
+      }
+
+      if (updateMap.size() != 0) {
+        updatePartColStatTable(updateMap, dbConn);
+      }
+
+      if (transactionalListeners != null) {
+        UpdatePartitionColumnStatEventBatch eventBatch = new UpdatePartitionColumnStatEventBatch(null);
+        for (Map.Entry entry : result.entrySet()) {
+          Map<String, String> parameters = (Map<String, String>) entry.getValue();
+          ColumnStatistics colStats = partColStatsMap.get(entry.getKey());
+          List<String> partVals = getPartValsFromName(tbl, colStats.getStatsDesc().getPartName());
+          UpdatePartitionColumnStatEvent event = new UpdatePartitionColumnStatEvent(colStats, partVals, parameters,
+                  tbl, writeId, null);
+          eventBatch.addPartColStatEvent(event);
+        }
+        MetaStoreListenerNotifier.notifyEventWithDirectSql(transactionalListeners,
+                EventMessage.EventType.UPDATE_PARTITION_COLUMN_STAT_BATCH, eventBatch, dbConn, sqlGenerator);
+      }
+      dbConn.commit();
+      committed = true;
+      return result;
+    } catch (Exception e) {
+      LOG.error("Unable to update Column stats for  " + tbl.getTableName(), e);
+      throw new MetaException("Unable to update Column stats for  " + tbl.getTableName()
+              + " due to: "  + e.getMessage());
+    } finally {
+      if (!committed) {
+        rollbackDBConn(dbConn);
+      }
+      closeDbConn(jdoConn);
+      unlockInternal();
+    }
+  }
+
+  /**
+   * Gets the next CS id from sequence MPartitionColumnStatistics and increment the CS id by numStats.
+   * @return The CD id before update.
+   */
+  public long getNextCSIdForMPartitionColumnStatistics(long numStats) throws MetaException {
+    Statement statement = null;
+    ResultSet rs = null;
+    long maxCsId = 0;
+    boolean committed = false;
+    Connection dbConn = null;
+    JDOConnection jdoConn = null;
+
+    try {
+      lockInternal();
+      jdoConn = pm.getDataStoreConnection();
+      dbConn = (Connection) (jdoConn.getNativeConnection());
+
+      setAnsiQuotes(dbConn);
+
+      // This loop will be iterated at max twice. If there is no records, it will first insert and then do a select.
+      // We are not using any upsert operations as select for update and then update is required to make sure that
+      // the caller gets a reserved range for CSId not used by any other thread.
+      boolean insertDone = false;
+      while (maxCsId == 0) {
+        String query = sqlGenerator.addForUpdateClause("SELECT \"NEXT_VAL\" FROM \"SEQUENCE_TABLE\" "
+                + "WHERE \"SEQUENCE_NAME\"= "
+                + quoteString("org.apache.hadoop.hive.metastore.model.MPartitionColumnStatistics"));
+        LOG.debug("Going to execute query " + query);
+        statement = dbConn.createStatement();
```

Comment:
where do we close the statement
see https://sonarcloud.io/project/issues?resolved=false&types=BUG&pullRequest=4517&id=apache_hive&open=AYvIijb_0txSvpyOT_4y

Reply:
no change here, copy of legacy code